File: artifacts/api-server/routers/caixa.py
"""Rotas de Caixa."""
import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from datetime import datetime
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from database import get_db
from models import SessaoCaixa, Sangria
from schemas import AbrirCaixaInput, SangriaInput

router = APIRouter()


def fmt_sessao(s: SessaoCaixa) -> dict:
    return {
        "id": s.id,
        "aberto_em": s.aberto_em.isoformat() if s.aberto_em else None,
        "fechado_em": s.fechado_em.isoformat() if s.fechado_em else None,
        "fundo_inicial": s.fundo_inicial,
        "total_dinheiro": s.total_dinheiro,
        "total_pix": s.total_pix,
        "total_cartao": s.total_cartao,
        "total_fiado": s.total_fiado,
        "total_sangria": s.total_sangria,
        "status": s.status,
    }


@router.get("/status")
def status_caixa(db: Session = Depends(get_db)):
    sessao = db.query(SessaoCaixa).filter(SessaoCaixa.status == "aberto").first()
    return {"aberto": sessao is not None, "sessao": fmt_sessao(sessao) if sessao else None}


@router.post("/abrir", status_code=201)
def abrir_caixa(data: AbrirCaixaInput, db: Session = Depends(get_db)):
    existente = db.query(SessaoCaixa).filter(SessaoCaixa.status == "aberto").first()
    if existente:
        raise HTTPException(status_code=400, detail="Já existe um caixa aberto")
    sessao = SessaoCaixa(fundo_inicial=data.fundo_inicial)
    db.add(sessao)
    db.commit()
    db.refresh(sessao)
    return fmt_sessao(sessao)


@router.post("/fechar")
def fechar_caixa(db: Session = Depends(get_db)):
    sessao = db.query(SessaoCaixa).filter(SessaoCaixa.status == "aberto").first()
    if not sessao:
        raise HTTPException(status_code=400, detail="Nenhum caixa aberto")
    sessao.status = "fechado"
    sessao.fechado_em = datetime.utcnow()
    db.commit()
    db.refresh(sessao)
    return fmt_sessao(sessao)


@router.post("/sangria", status_code=201)
def registrar_sangria(data: SangriaInput, db: Session = Depends(get_db)):
    sessao = db.query(SessaoCaixa).filter(SessaoCaixa.status == "aberto").first()
    if not sessao:
        raise HTTPException(status_code=400, detail="Nenhum caixa aberto")
    sangria = Sangria(sessao_id=sessao.id, valor=data.valor, motivo=data.motivo)
    db.add(sangria)
    sessao.total_sangria += data.valor
    db.commit()
    db.refresh(sangria)
    return {
        "id": sangria.id,
        "sessao_id": sangria.sessao_id,
        "valor": sangria.valor,
        "motivo": sangria.motivo,
        "criado_em": sangria.criado_em.isoformat(),
    }


@router.get("/sangrias")
def listar_sangrias(sessao_id: Optional[int] = None, db: Session = Depends(get_db)):
    query = db.query(Sangria)
    if sessao_id:
        query = query.filter(Sangria.sessao_id == sessao_id)
    sangrias = query.order_by(Sangria.criado_em.desc()).all()
    return [
        {
            "id": s.id,
            "sessao_id": s.sessao_id,
            "valor": s.valor,
            "motivo": s.motivo,
            "criado_em": s.criado_em.isoformat(),
        }
        for s in sangrias
    ]


@router.get("/historico")
def historico_caixa(db: Session = Depends(get_db)):
    sessoes = db.query(SessaoCaixa).order_by(SessaoCaixa.aberto_em.desc()).limit(30).all()
    return [fmt_sessao(s) for s in sessoes]

File: artifacts/api-server/src/routes/caixa.ts
import { Router, type IRouter } from "express";
import { db } from "@workspace/db";
import { sessoesCaixaTable, sangriasTable, vendasTable } from "@workspace/db/schema";
import { eq, desc } from "drizzle-orm";
import { AbrirCaixaBody, RegistrarSangriaBody } from "@workspace/api-zod";
import { sql } from "drizzle-orm";
import ExcelJS from "exceljs";
import { itensVendaTable } from "@workspace/db/schema";

const router: IRouter = Router();

function formatSessao(s: typeof sessoesCaixaTable.$inferSelect) {
  return {
    ...s,
    aberto_em: s.aberto_em.toISOString(),
    fechado_em: s.fechado_em?.toISOString() ?? null,
  };
}

router.get("/:sessaoId/relatorio.xlsx", async (req, res) => {
  const sessaoId = Number(req.params.sessaoId);
  const [sessao] = await db.select().from(sessoesCaixaTable).where(eq(sessoesCaixaTable.id, sessaoId));
  if (!sessao) return res.status(404).json({ ok: false, message: "Sessão não encontrada" });

  const vendas = await db.select().from(vendasTable).where(eq(vendasTable.sessao_id, sessaoId));
  const sangriasSessao = await db.select().from(sangriasTable).where(eq(sangriasTable.sessao_id, sessaoId));

  const wb = new ExcelJS.Workbook();

  const resumo = wb.addWorksheet("Resumo");
  resumo.addRows([
    ["Sessão", sessaoId],
    ["Abertura", sessao.aberto_em.toLocaleString("pt-BR")],
    ["Fechamento", sessao.fechado_em?.toLocaleString("pt-BR") ?? "Aberta"],
    ["Fundo inicial", sessao.fundo_inicial],
    ["Dinheiro", sessao.total_dinheiro],
    ["PIX", sessao.total_pix],
    ["Cartão", sessao.total_cartao],
    ["Fiado", sessao.total_fiado],
    ["Sangrias", sessao.total_sangria],
  ]);

  const abaVendas = wb.addWorksheet("Vendas");
  abaVendas.columns = [
    { header: "ID", key: "id", width: 8 },
    { header: "Hora", key: "hora", width: 20 },
    { header: "Categoria", key: "categoria", width: 12 },
    { header: "Pagamento", key: "pagamento", width: 12 },
    { header: "Total", key: "total", width: 12 },
    { header: "Desconto", key: "desconto", width: 10 },
  ];
  abaVendas.addRows(vendas.map(v => ({
    id: v.id, hora: v.criado_em.toLocaleString("pt-BR"),
    categoria: v.categoria, pagamento: v.pagamento, total: v.total, desconto: v.desconto,
  })));

  const abaSangrias = wb.addWorksheet("Sangrias");
  abaSangrias.columns = [
    { header: "Hora", key: "hora", width: 20 },
    { header: "Valor", key: "valor", width: 12 },
    { header: "Motivo", key: "motivo", width: 30 },
  ];
  abaSangrias.addRows(sangriasSessao.map(s => ({
    hora: s.criado_em.toLocaleString("pt-BR"), valor: s.valor, motivo: s.motivo ?? "",
  })));

  res.setHeader("Content-Type", "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
  res.setHeader("Content-Disposition", `attachment; filename=caixa-${sessaoId}.xlsx`);
  await wb.xlsx.write(res);
  res.end();
});

router.get("/status", async (_req, res) => {
  const [sessao] = await db
    .select()
    .from(sessoesCaixaTable)
    .where(eq(sessoesCaixaTable.status, "aberto"))
    .limit(1);

  res.json({
    aberto: !!sessao,
    sessao: sessao ? formatSessao(sessao) : null,
  });
});

router.post("/abrir", async (req, res) => {
  const [existente] = await db
    .select()
    .from(sessoesCaixaTable)
    .where(eq(sessoesCaixaTable.status, "aberto"))
    .limit(1);

  if (existente) {
    res.status(400).json({ ok: false, message: "Já existe um caixa aberto" });
    return;
  }

  const data = AbrirCaixaBody.parse(req.body);
  const [sessao] = await db
    .insert(sessoesCaixaTable)
    .values({ fundo_inicial: data.fundo_inicial ?? 0 })
    .returning();

  res.status(201).json(formatSessao(sessao));
});

router.post("/fechar", async (_req, res) => {
  const [sessao] = await db
    .select()
    .from(sessoesCaixaTable)
    .where(eq(sessoesCaixaTable.status, "aberto"))
    .limit(1);

  if (!sessao) {
    res.status(400).json({ ok: false, message: "Nenhum caixa aberto" });
    return;
  }

  const [fechada] = await db
    .update(sessoesCaixaTable)
    .set({ status: "fechado", fechado_em: new Date() })
    .where(eq(sessoesCaixaTable.id, sessao.id))
    .returning();

  res.json(formatSessao(fechada));
});

router.post("/sangria", async (req, res) => {
  const [sessao] = await db
    .select()
    .from(sessoesCaixaTable)
    .where(eq(sessoesCaixaTable.status, "aberto"))
    .limit(1);

  if (!sessao) {
    res.status(400).json({ ok: false, message: "Nenhum caixa aberto" });
    return;
  }

  const data = RegistrarSangriaBody.parse(req.body);

  const [sangria] = await db
    .insert(sangriasTable)
    .values({
      sessao_id: sessao.id,
      valor: data.valor,
      motivo: data.motivo ?? null,
    })
    .returning();

  await db
    .update(sessoesCaixaTable)
    .set({ total_sangria: sql`total_sangria + ${data.valor}` })
    .where(eq(sessoesCaixaTable.id, sessao.id));

  res.status(201).json({
    ...sangria,
    criado_em: sangria.criado_em.toISOString(),
  });
});

router.get("/sangrias", async (req, res) => {
  const { sessao_id } = req.query as { sessao_id?: string };
  const condition = sessao_id
    ? eq(sangriasTable.sessao_id, parseInt(sessao_id))
    : undefined;

  const sangrias = await db
    .select()
    .from(sangriasTable)
    .where(condition)
    .orderBy(desc(sangriasTable.criado_em));

  res.json(
    sangrias.map((s) => ({
      ...s,
      criado_em: s.criado_em.toISOString(),
    }))
  );
});

router.get("/historico", async (_req, res) => {
  const sessoes = await db
    .select()
    .from(sessoesCaixaTable)
    .orderBy(desc(sessoesCaixaTable.aberto_em))
    .limit(30);

  res.json(sessoes.map(formatSessao));
});

export default router;
